File: job-2022-autumn/486.predict-the-winner.js
/*
 * @lc app=leetcode id=486 lang=javascript
 *
 * [486] Predict the Winner
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var PredictTheWinner = function (nums) {
  /**
   * dp[i][j] 取 [i,j] 区间的石头，最多能比对方多拿多少分
   *
   * dp[i][i] = nums[i]
   *
   * dp[i][j] = Math.max(
   *  nums[i] - dp[i+1][j],
   *  nums[j] - dp[i][j-1]
   * )
   */

  /** @type {number[]} */
  const dp = new Array(nums.length);

  for (let i = nums.length - 1; i >= 0; i--) {
    dp[i] = nums[i];
    for (let j = i + 1; j < nums.length; j++) {
      dp[j] = Math.max(nums[i] - dp[j], nums[j] - dp[j - 1]);
    }
  }

  return dp[nums.length - 1] >= 0;
};
// @lc code=end

File: src/2026/486.predict-the-winner.cpp
/*
 * @lc app=leetcode id=486 lang=cpp
 *
 * [486] Predict the Winner
 */

#include <optional>
#include <vector>

// @lc code=start
class Solution {
 public:
  bool predictTheWinner(const std::vector<int>& nums) {
    std::vector<std::vector<std::optional<int>>> cache(
        nums.size(), std::vector<std::optional<int>>(nums.size()));
    return GetMaxScoreDiff(nums, 0, nums.size() - 1, cache) >= 0;
  }

 private:
  // If current player can now first pick from nums[left...right], the max score
  // diff between current player and opponent.
  static int GetMaxScoreDiff(
      const std::vector<int>& nums,
      const int left,
      const int right,
      std::vector<std::vector<std::optional<int>>>& cache) {
    if (left == right) {
      return nums[left];
    }
    if (cache[left][right].has_value()) {
      return cache[left][right].value();
    }

    const int max_score_diff =
        std::max(nums[left] - GetMaxScoreDiff(nums, left + 1, right, cache),
                 nums[right] - GetMaxScoreDiff(nums, left, right - 1, cache));
    cache[left][right] = max_score_diff;
    return max_score_diff;
  }
};
// @lc code=end
